File: examples/awfy/go/permute.go
package main

var permuteV []int
var permuteCount int

func permuteSwap(i, j int) {
	permuteV[i], permuteV[j] = permuteV[j], permuteV[i]
}

func permuteRec(n int) {
	permuteCount++
	if n != 0 {
		permuteRec(n - 1)
		for i := n - 1; i >= 0; i-- {
			permuteSwap(n-1, i)
			permuteRec(n - 1)
			permuteSwap(n-1, i)
		}
	}
}

func permuteRun(size int) int {
	for rep := 0; rep < size; rep++ {
		permuteCount = 0
		permuteV = make([]int, 6)
		permuteRec(6)
	}
	return permuteCount
}

var permuteBench = Bench{Name: "permute", Warmup: 10, Iters: 40, Size: 1000, Expected: 8660, Run: permuteRun}

File: examples/awfy/node/permute.mjs
export const warmup = 10, iters = 40, size = 1000, expected = 8660;

let v;
let count;

function permute(n) {
  count++;
  if (n !== 0) {
    permute(n - 1);
    for (let i = n - 1; i >= 0; i--) {
      swap(n - 1, i);
      permute(n - 1);
      swap(n - 1, i);
    }
  }
}

function swap(i, j) {
  const tmp = v[i];
  v[i] = v[j];
  v[j] = tmp;
}

export function run(size) {
  for (let rep = 0; rep < size; rep++) {
    count = 0;
    v = new Array(6).fill(0);
    permute(6);
  }
  return count;
}
